src/hardware/bk-1856d/api.c

```diff
@@ -0,0 +1,247 @@
+/*
+ * This file is part of the libsigrok project.
+ *
+ * Copyright (C) 2020 LUMERIIX/danselmi <.>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <config.h>
+#include "protocol.h"
+
+#define SERIALCOMM "9600/8n1/dtr=1/rts=0"
+
+static const uint32_t scanopts[] = {
+	SR_CONF_CONN,
+};
+
+static const uint32_t drvopts[] = {
+	SR_CONF_MULTIMETER,
+};
+
+static const uint32_t devopts[] = {
+	SR_CONF_CONTINUOUS,
+	/** using TIMEBASE to select the gate time */
+	SR_CONF_TIMEBASE      | SR_CONF_SET |SR_CONF_GET| SR_CONF_LIST,
+	SR_CONF_LIMIT_SAMPLES | SR_CONF_SET,
+	SR_CONF_DATA_SOURCE   | SR_CONF_GET | SR_CONF_SET | SR_CONF_LIST,
+};
+
+static struct sr_dev_driver bk_1856d_driver_info;
+
+const uint64_t timebases[][2] = {
+	/*miliseconds*/
+	{ 10, 1000 },
+	{ 100, 1000 },
+	/* seconds */
+	{ 1, 1 },
+	{ 10, 1 },
+};
+
+static const char *data_sources[] = {
+	"Input A", "Input C",
+};
+
+static GSList *scan(struct sr_dev_driver *di, GSList *options)
+{
+	struct dev_context *devc;
+	struct sr_dev_inst *sdi;
+	GSList *opt;
+	const char *conn;
+
+	conn = NULL;
+	for (opt = options; opt; opt = opt->next) {
+		struct sr_config *src = opt->data;
+		switch (src->key) {
+		case SR_CONF_CONN:
+			conn = g_variant_get_string(src->data, NULL);
+			break;
+		}
+	}
+	if (!conn)
+		return NULL;
+
+	sdi = g_malloc0(sizeof(struct sr_dev_inst));
+	sdi->status = SR_ST_INACTIVE;
+	sdi->vendor = g_strdup("BK Precision");
+	sdi->model = g_strdup("bk-1856d");
+	devc = g_malloc0(sizeof(struct dev_context));
+	sr_sw_limits_init(&(devc->sw_limits));
+	sdi->conn = sr_serial_dev_inst_new(conn, SERIALCOMM);
+	sdi->inst_type = SR_INST_SERIAL;
+	sdi->priv = devc;
+	g_mutex_init(&devc->rw_mutex);
+	sr_channel_new(sdi, 0, SR_CHANNEL_ANALOG, TRUE, "P1");
+	devc->sel_input = InputC;
+	devc->curr_sel_input = InputC;
+	devc->gate_time = 0;
+	devc->hold = Off;
+
+	g_usleep(150 * 1000); /* Wait a little to allow serial port to settle. */
```

Comment:
I found no other solution than this (same for the sleep time after changing the gate time).
If I remember correctly (it's more than two years since), I found the values empirically and rounded up.

Reply:
But you haven't actually sent anything to the device yet, `devc->gate_time` is just a bit of memory. Why would you need a delay?